module-system/module-implementations/sov-nft-module/src/collection.rs

```diff
@@ -0,0 +1,134 @@
+use anyhow::anyhow;
+use sov_modules_api::{Context, StateMap, WorkingSet};
+
+use crate::address::CollectionAddress;
+use crate::utils::get_collection_address;
+use crate::CreatorAddress;
+
+#[cfg_attr(
+    feature = "native",
+    derive(serde::Serialize),
+    derive(serde::Deserialize)
+)]
+#[derive(borsh::BorshDeserialize, borsh::BorshSerialize, Debug, PartialEq, Clone)]
+/// Defines an nft collection
+pub struct Collection<C: Context> {
+    /// Name of the collection
+    /// The name has to be unique in the scope of a creator. A single creator address cannot have
+    /// duplicate collection names
+    name: String,
+    /// Address of the collection creator
+    /// This is the only address that can mint new NFTs for the collection
+    creator: CreatorAddress<C>,
+    /// If a collection is frozen, then new NFTs
+    /// cannot be minted and the supply is frozen
+    frozen: bool,
+    /// Supply of the collection. This is dynamic and changes
+    /// with the number of NFTs created. It stops changing
+    /// when frozen is set to true.
+    supply: u64,
+    /// collection metadata stored at this url
+    collection_uri: String,
+}
+
+pub enum CollectionState<C: Context> {
+    Frozen(Collection<C>),
+    Mutable(MutableCollection<C>),
+}
+
+impl<C: Context> Collection<C> {
+    pub fn new(
+        collection_name: &str,
+        collection_uri: &str,
+        collections: &StateMap<CollectionAddress<C>, Collection<C>>,
+        context: &C,
+        working_set: &mut WorkingSet<C>,
+    ) -> anyhow::Result<(CollectionAddress<C>, Collection<C>)> {
+        let creator = context.sender();
+        let ca = get_collection_address(collection_name, creator.as_ref());
+        let c = collections.get(&ca, working_set);
+        if c.is_some() {
+            Err(anyhow!(
+                "Collection with name: {} already exists creator {}",
+                collection_name,
+                creator
+            ))
+        } else {
+            Ok((
+                ca,
+                Collection {
+                    name: collection_name.to_string(),
+                    creator: CreatorAddress::new(creator),
+                    frozen: false,
+                    supply: 0,
+                    collection_uri: collection_uri.to_string(),
+                },
+            ))
+        }
+    }
+
+    pub fn get_owned_collection(
+        collection_name: &str,
+        collections: &StateMap<CollectionAddress<C>, Collection<C>>,
+        context: &C,
+        working_set: &mut WorkingSet<C>,
+    ) -> anyhow::Result<(CollectionAddress<C>, CollectionState<C>)> {
+        let creator = context.sender();
+        let ca = get_collection_address(collection_name, creator.as_ref());
+        let c = collections.get(&ca, working_set);
```

Comment:
nit: let's use full variable names here. Even `addr` is clear enough, but one to two letters is a little short :)